### PR TITLE
ARROW-9555: [Rust] [DataFusion] Implement physical node for inner join

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR is based on top of #8630 and contains a physical node to perform an inner join in DataFusion.

This is still a draft, but IMO the design is here and the two tests already pass.

This is co-authored with @andygrove , that contributed to the design on how to perform this operation in the context of DataFusion (see ARROW-9555 for details).

The API used for the computation of the join at the arrow level is briefly discussed in [this document](https://docs.google.com/document/d/1KKuBvfx7uKi-x-tWOL60R1FNjDW8B790zPAv6yAlYcU/edit).

There is still a lot to work on, but I I though it would be a good time to have a first round of discussions, and also to gauge timings wrt to the 3.0 release.

There are two main issues being addressed in this PR:

* How to we perform the join at the partition level: this pr collects all batches from the left, and then issues a stream per part on the right. Each batch on that stream joins itself with all the ones from the left (N) via a hash. This allow us to only require computing the hash of a row once (first all the left, then one by one on the right).

* How do we build an array from `N (left)` arrays and a set of indices (matching the hash from the right): this is done using the `MutableArrayData` being worked on #8630, which incrementally memcopies slots from each of the N arrays based on the index. This implementation is useful because it works for all array types and does not require casting anything to rust native types (i.e. it operates on `ArrayData`, not specific implementations).

There are still some steps left to have a join in SQL, most notably the whole logical planning, the output_partition logic, the bindings to SQL and DataFrame API, update the optimizers to handle nodes with 2 children,  and a whole battery of tests.

There is also a natural path for the other joins, as it will be a matter of incorporating the work already on PR #8689 that introduces the option to extend the `MutableArrayData` with nulls, the operation required for left and right joins.

### CODE REVIEW

File: rust/datafusion/src/physical_plan/hash_utils.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+//! Functionality used both on logical and physical plans
+
+use crate::error::{DataFusionError, Result};
+use arrow::datatypes::{Field, Schema};
+use std::collections::HashSet;
+
+/// All valid types of joins.
+#[derive(Clone, Debug)]
+pub enum JoinHow {
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/hash_utils.rs:26]: nit: `JoinType` would be a more standard name

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+/// join execution plan executes partitions in parallel and combines them into a set of
+/// partitions.
+#[derive(Debug)]
+pub struct HashJoinExec {
+    /// left side
+    left: Arc<dyn ExecutionPlan>,
+    /// right side
+    right: Arc<dyn ExecutionPlan>,
+    /// Set of common columns used to join on
+    on: HashSet<String>,
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/hash_join.rs:63]: This is a great start but will limit us to cases where the column names are the same between two tables. As a follow-on we can expand this to support different names e..g `customer.id = orders.customer_id`.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+use crate::error::{DataFusionError, Result};
+
+use super::{
+    group_scalar::GroupByScalar, ExecutionPlan, Partitioning, RecordBatchStream,
+    SendableRecordBatchStream,
+};
+
+// An index of (batch, row) uniquely identifying a row in a part.
+type Index = (usize, usize);
+// None represents a null (e.g. in case of a left join, some right indices are null)
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_join.rs:46]: this comment is out-of-sync.

Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_join.rs:46]:
```suggestion
// A pair (left index, right index)
```

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+        }
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.right.output_partitioning()
+    }
+
+    async fn execute(&self, partition: usize) -> Result<SendableRecordBatchStream> {
+        // build the keys
+        let stream = self.left.execute(partition).await?;
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_join.rs:133]: this is wrong: we need to get all partitions from the left here.

Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_join.rs:133]: @andygrove , I understand that the idea is to build a vector of RecordBatches from the left once, and then use it on the multiple parts on the right. However, I do not think we have a place to store it atm, as this is information that is required by multiple parts from the right, and our API does not support mutating `self` on `execute`, which means that we can't "save" and share this vector across the right partition.

 if we perform this operation on `execute`, we need to compute it once per right part. If we perform this operation on `new`, we can no longer plan without hitting the data.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.right.output_partitioning()
+    }
+
+    async fn execute(&self, partition: usize) -> Result<SendableRecordBatchStream> {
+        // merge all parts into a single stream
+        // this is currently expensive as we re-compute this for every part from the right
+        // TODO: Fix this issue: we can't share this state across parts on the right.
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/hash_join.rs:146]: When I tried to implement this, my approach was to store the materialized left-hand side in an `Arc<Mutex<_>>` and my plan was to use logic like this:

```rust
if (partition == 0) {
  // materialize left input and store in Arc<Mutex<_>>
} else {
 // wait for left input to be materialized
}
```

Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_join.rs:146]: I woulld be fine with us doing this as a follow up and get the expensive version of this functionally correct before we optimize too much.

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_join.rs:68]: After thinking about this some more, I think it might be better to address this in the current PR. I would suggest either:

```rust
left_keys: Vec<String>
right_keys: Vec<String>,
```

or the more general case:

```rust
left_keys: Vec<Arc<dyn PhysicalExpr>>
right_keys: Vec<Arc<dyn PhysicalExpr>>,
```

It should be possible to perform an equi-join using any deterministic expression, for example:

```
SELECT a.*, b.* FROM a JOIN b on UPPER(a.name) = UPPER(CONCAT(b.first, ' ', b.last))
```

File: rust/datafusion/src/physical_plan/hash_utils.rs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            let left_fields = left.fields().iter().filter(|f| !on.contains(f.name()));
+
+            let right_fields = right.fields().iter().filter(|f| !on.contains(f.name()));
+
+            // "on" are first by construction, then left, then right
+            on_fields
+                .chain(left_fields)
+                .chain(right_fields)
+                .map(|f| f.clone())
+                .collect()
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/hash_utils.rs:98]: It would be more standard to have all the left fields followed by all the right fields. Also, the new schema should have fully qualified names that include the table/alias prefix.

For example if we join table 'a' with fields `id` and name` with table 'b' that also has fields `id` and `name`, the output schema would be `a.id`, `a.name`, `b.id`, `b.name`.

It is quite possible that we are lacking support currently for being able to reference these qualified field names, so we will need to deal with that before we can fully integrate this new join operator into DataFusion.

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+                &self.join_type,
+            )?)),
+            _ => Err(DataFusionError::Internal(
+                "HashJoinExec wrong number of children".to_string(),
+            )),
+        }
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.right.output_partitioning()
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/hash_join.rs:140]: This is correct. The output partitioning is the same as the partitioning of the right-hand input.

File: rust/datafusion/src/physical_plan/mod.rs [hunk 8]
```diff
@@ -89,6 +89,8 @@ pub trait ExecutionPlan: Debug + Send + Sync {
 pub enum Partitioning {
     /// Unknown partitioning scheme
     UnknownPartitioning(usize),
+    /// parts hashed by columns and `N` partitions
+    HashPartitioning(Vec<String>, usize),
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/mod.rs:93]: It makes sense for us to add a `HashPartitioning` variant here, but this isn't required for implementing the hash join operator since we aren't changing partitioning.

**Comment by andygrove** [comment]
I did a first pass review and this is looking really great @jorgecarleitao. I left comments for the things that I think need to be addressed in this PR. I will be able to help out with plumbing this through from SQL/DataFrame and query planner. This is going to allow us to support most of the TPC-H queries so I am really excited to see this get merged :rocket:

File: rust/datafusion/src/physical_plan/mod.rs (hunk 8 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/mod.rs:93]: You are right, I was working on the repartition and join at the same time at some point, and this slipped through the cracks.

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_join.rs:68]: Agreed. I will have this fixed: I do not see any technical challenge here (apart from the schema changes).

File: rust/datafusion/src/physical_plan/hash_utils.rs (hunk 6 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: Well spotted.

Since expressions may now contain columns from multiple Schemas/RecordBatch, I think that we need to modify all signatures of `PhysicalExpr` to receive multiple `Schema` and `RecordBatch`, as expressions may now contain columns from more than one `RecordBatch`.

I agree that this is a blocker. I will work on that and leave this PR for now, as we can't progress any further here.

**Comment by jorgecarleitao** [comment]
> I did a first pass review and this is looking really great @jorgecarleitao. I left comments for the things that I think need to be addressed in this PR. I will be able to help out with plumbing this through from SQL/DataFrame and query planner. This is going to allow us to support most of the TPC-H queries so I am really excited to see this get merged 🚀

Thanks a lot, @andygrove . I will be focusing on the physical plans alone, so any help on the SQL/DataFrame and logical plans would be really appreciated.

File: rust/datafusion/src/physical_plan/hash_utils.rs (hunk 6 above)
Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: I don't understand why we would want to do that. Each operator has a single output schema. The join operator creates a new schema derived from the input schemas.

Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: The output schema of the join would simply have fields named "a.id", "a.name", "b.id", "b.name" based on the previous example.

Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: @jorgecarleitao Just wanted to make sure you saw this :point_up:

Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: You are right. I am not thinking strait. 

My (incorrect) concern was that expressions of the form `UPPER(a.name) = UPPER(CONCAT(b.first, ' ', b.last))` would need to be represented as `PhysicalExpr`, but I totally forgot that the `=` there is _not_ the equal operator, but the "when equal hash", and therefore that expression does not mean "left equals right", that a physical expression would encapsulate.

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_join.rs:68]: Another option here is to have an optimizer rule that whenever a non-column expression (e.g. `a.a + 1 = a.b + 2`) is passed to the `on`, it converts it to a projection on both sides. In terms of compute, AFAIK it is the same, as we always need to evaluate the array before computing the hash of each of its rows, and evaluating it on the `Join` or on a preceding projection seems to be equivalent.

The rational for using `HashSet<String>` was that this is way easier to represent (i.e. serialize and equate) on a partition key, which means when we introduce hashJoins on partitioned keys, it may be easier to tackle.

File: rust/datafusion/src/physical_plan/hash_utils.rs (hunk 6 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: > For example if we join table 'a' with fields id and name with table 'b' that also has fields id and name, the output schema would be a.id, a.name, b.id, b.name.

In general the nodes are not named, which means that `'a'` is arbitrary. Spark DataFrames do not append the qualifiers, as nodes are usually not named (`df.select('a', 'b').join(df1.select('a', 'c'), on='a')` returns columns `'a','b','c'`). 

What I dislike in Spark is that it allows two columns to have the same name, i.e. `df.select('a', 'b').join(df1.select('a', 'b'), on='a')` returns columns `'a','b','b'`, but if we then write `.select('b')`, it raises.

The current implementation uses what I consider an ideal implementation: no qualifiers are added, but it raises during the planning of the join if two columns will end up with the same name (as we do for other nodes). I.e. this would be possible:

```rust
df
    .select('a', 'b')
    .join(df1, on=col("a", 0) == col("a", 1))
    .select(vec![col('a')])
```

If we introduce qualifiers, we need to use something like

```rust
df
    .select('a', 'b')
    .join(df1, on=on_col("a", 0) == on_col("a", 1))
    .select(vec![col('?.b')])
```

where `?` is an arbitrary qualifier (`1?`, `t1`? `a`?). Moreover, that qualifier will depend on whether `df1` has a column named `b` or not, which IMO leads to surprising results (the resulting schema may be `?.b` or `b` depending on whether the nodes have colliding column names).

The current code implements the principle of least surprise together with the "explicit is better than implicit": columns are named according to what the user set them to be named, and it is the user's responsibility to ensure that two column names do not collide in a join (i.e. DataFusion errors with a nice message).

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_join.rs:68]: Yes, this is an implicit projection that happens on the join expressions. This is usually not represented as a separate step in a query plan, in my experience at least, and is handled within the join operator itself. I agree that this would be equivalent in compute to having a separate projection in the plan.
 
I would be fine if we just want to support join on columns for this PR and look at supporting any expression as a separate issue. We can always work around the limitation by doing explicit projections before the join.

File: rust/datafusion/src/physical_plan/hash_utils.rs (hunk 6 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: Note that I am looking at this from the optic of the DataFrame API. For SQL I imagine that `a.b` is prevalent because that is how in SQL we refer to a column of a specific table `a` (as all tables are named).

Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: I think you're right. This is simpler and forces the user to add aliases as required to avoid conflicts.

Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: I do think we need to support the case where the join key columns have different names in each table though. Here is an example from one of the simplest TPC-H queries.

```sql
FROM
    orders,
    lineitem
WHERE
    o_orderkey = l_orderkey
```

Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: Good point. I will prototype to double check that it works well at the schema validation.

Just to check whether we agree on some cases:

```
df.columns = ['a', 'l_orderkey']

df1.columns = ['o_orderkey', 'l_orderkey']

df.join(df1, on=df['l_orderkey'] == df1['o_orderkey'])

# raises an error (there will be two columns named `l_orderkey`).
```

```
df.columns = ['l_orderkey', 'b']

df1.columns = ['o_orderkey', 'c']

df.join(df1, on=df['l_orderkey'] == df1['o_orderkey'])

# Ok: columns = ['l_orderkey', 'b', 'o_orderkey', 'c']  (order: left, right, different column names, both are returned)
```

```
df.columns = ['a', 'o_orderkey']

df1.columns = ['o_orderkey', 'c']

df.join(df1, on=df['o_orderkey'] == df1['o_orderkey'])

# Ok: columns = ['a', 'o_orderkey', 'c']  (order: left, right, same column name, only one is returned)
```

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (alamb) [reply rust/datafusion/src/physical_plan/hash_join.rs:68]: For joins, the `column`=`column` case, the so called `equijoin` is so common and critical, I think it is important to make *super fast* and thus special cased. The more general expression case, while supported, is much less important than fast equi-joins, I think

I have seen both approaches (the join handles generic expressions, or the planner arranges to have only equijoins). For OUTER joins there is no choice but to do the evaluation in the join operator itself

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 5 above)
Comment (alamb) [reply rust/datafusion/src/physical_plan/hash_join.rs:146]: I agree with this suggestion

File: rust/datafusion/src/physical_plan/hash_utils.rs (hunk 6 above)
Comment (alamb) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: I don't think we should error (long term) if the two inputs have the same column names. It is fine for an initial  implementation, but I think we should automatically created aliases to disambiguate such outputs long term

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+// E.g. [1, 2] -> [(0, 3), (1, 6), (0, 8)] indicates that (column1, column2) = [1, 2] is true
+// for rows 3 and 8 from batch 0 and row 6 from batch 1.
+type JoinHashMap = HashMap<Vec<GroupByScalar>, Vec<Index>>;
+type JoinLeftData = (JoinHashMap, Vec<RecordBatch>);
+
+/// join execution plan executes partitions in parallel and combines them into a set of
+/// partitions.
+#[derive(Debug)]
+pub struct HashJoinExec {
+    /// left side
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_join.rs:63]: "left, build side  -- built into a hash table"

Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_join.rs:63]: "/// left (build) side which gets hashed"

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+    left: Arc<dyn ExecutionPlan>,
+    /// right side
+    right: Arc<dyn ExecutionPlan>,
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_join.rs:66]: "/// right (probe) side which are filtered by the hash table"

File: rust/datafusion/src/physical_plan/hash_join.rs [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+                let hash = &mut acc.0;
+                let values = &mut acc.1;
+                let index = acc.2;
+                update_hash(&self.on, &batch, hash, index).unwrap();
+                values.push(batch);
+                acc.2 += 1;
+                Ok(acc)
+            })
+            .await?;
+        // we have the batches and the hash map with their keys. We can how create a stream
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_join.rs:166]: "how --> now"

File: rust/datafusion/src/physical_plan/hash_utils.rs [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    /// Inner join
+    Inner,
+}
+
+/// Checks whether the schemas "left" and "right" and columns "on" represent a valid join.
+/// They are valid whenever their columns' intersection equals the set `on`
+pub fn check_join_is_valid(
+    left: &Schema,
+    right: &Schema,
+    on: &HashSet<String>,
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_utils.rs:36]: This might be easier code to write if the `on` expresion was of the form:

```
equality_expressions: Vec<(Expr, Expr)>
other_expressions: Vec<Expr>
```

Where the planner identifies quality expressions and breaks them up -- again I can help write this code.

**Review by alamb — APPROVED** [review]
TLDR; Thank you very much @jorgecarleitao  ❤️ ❤️  -- I haven't reviewed all this code carefully, but at this point given how large this PR is and how important this functionality is, I suggest we get this PR merged in as quickly as possible and then iterate on it afterwards. 

Joins in particular can get arbitrarily complicated I think, so just getting something in and iterating is key.

I suspect as we start plumbing joins through the planner we'll find things we want to change, but I don't think we are likely to get all that right if we try to guess it up front. 

Some other thoughts:

* Joins typically use the terms "inner" and "outer" for their two input relations. I would describe this operator as a classic "inner join" where the semantics of the two inputs are identical (though performance characteristics are not)

* In a hash join such as this, terms from the literature are "build" input (the side that you build the the hash table from) and the probe input (the side you test for being in the hash table)

FWIW I am used to seeing the build side on the right, and the probe side on the left, but that is just convention.

> this pr collects all batches from the left, and then issues a stream per part on the right.

This is the standard hash join approach and I think it is a good initial implementation. There are more sophisticated strategies that re-partition both the left and right inputs so you can build multiple disjoint hashtables, but that doesn't always go faster and is significantly more complicated to implement

> most notably the whole logical planning, the output_partition logic, the bindings to SQL and DataFrame API

I can make time ot help with this -- it is directly related to some of the longer term thing we want to do in IOx and I have some experience in doing it.

File: rust/datafusion/src/physical_plan/hash_utils.rs (hunk 6 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_utils.rs:98]: @alamb , it is difficult to argue against no arguments :)

My argument is based on the principle of least surprise: every time we desambiguate, we surprise the users by mutating the finely crafted plan that they created.

Note that this does not happen in SQL, as in SQL all names are disambiguated by the table qualifier (e.g. `SELECT a.b, b.b ...`)

**Comment by jorgecarleitao** [comment]
Hi, thanks a lot for the feedback @andygrove and @alamb . I have changed this PR in the following ways:

1. removed all changes wrt to the `filter` and just kept the `MutableDataArray` construct for the join, to not mix the changes.
2. Made "on" be `Vec<(String, String)>`.

The tests are not passing, but I will work on them over the weekend.

File: rust/datafusion/src/physical_plan/hash_utils.rs [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+            let right_fields = right
+                .fields()
+                .iter()
+                .filter(|f| !on_right.contains(f.name()));
+
+            // left then right
+            left_fields.chain(right_fields).map(|f| f.clone()).collect()
+        }
+    };
+    Schema::new(fields)
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/hash_utils.rs:98]: @jorgecarleitao I think the join keys are being dropped from the schema?

File: rust/datafusion/src/physical_plan/hash_utils.rs [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    right: &HashSet<String>,
+    on: &JoinOn,
+) -> Result<()> {
+    if on.len() == 0 {
+        return Err(DataFusionError::Plan(
+            "The 'on' clause of a join cannot be empty".to_string(),
+        ));
+    }
+    let on_left = &on.iter().map(|on| on.0.to_string()).collect::<HashSet<_>>();
+    let left_missing = left.difference(on_left).collect::<HashSet<_>>();
```
Comment (andygrove) [inline rust/datafusion/src/physical_plan/hash_utils.rs:57]: I'm confused here and I think this code might not be doing what you intended. `left_missing` contains all the left fields except for the fields references in the `on` clause, so this function will always fail unless the both tables only have join keys and no other columns.

Comment (andygrove) [reply rust/datafusion/src/physical_plan/hash_utils.rs:57]: Most of the tests pass with this code:

```rust
    let on_left = &on.iter().map(|on| on.0.to_string()).collect::<HashSet<_>>();
    let left_keys = left
        .iter()
        .filter(|f| on_left.contains(*f))
        .map(|s| s.clone())
        .collect::<HashSet<_>>();
    let left_missing = on_left.difference(&left_keys).collect::<HashSet<_>>();

    let on_right = &on.iter().map(|on| on.1.to_string()).collect::<HashSet<_>>();
    let right_keys = right
        .iter()
        .filter(|f| on_right.contains(*f))
        .map(|s| s.clone())
        .collect::<HashSet<_>>();
    let right_missing = on_right.difference(&right_keys).collect::<HashSet<_>>();

    if left_missing.len() > 0 || right_missing.len() > 0 {
        return Err(DataFusionError::Plan(format!(
                "The left or right side of the join does not have columns \"on\": \nMissing on the left: {:?}\nMissing on the right: {:?}",
                left_missing,
                right_missing,
            )));
    };
```

Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_utils.rs]: I am sorry for the confusion: pushed without testing 🤦

I think that this is ready now for the second round. Main changes are described above.

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 5 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_join.rs:153]: I will investigate this `execute` limitation. I have an idea about it: I think that we should change `async execute` to return a vector of streams (one stream per part). I.e. instead of 

```rust
let stream_0 = node.execute(0)?.await;
let stream_1 = node.execute(1)?.await;
...
```

we use 

```rust
let streams = node.execute()?.await;
let stream_0 = streams[0];
let stream_1 = streams[1];
...
```

by doing this, we can make `node.execute()`'s implementation to compute the left side, and have that shared across all streams (`HashJoinStream`) without mutating the node (which IMO is good that it is immutable).

IMO we may not need an `Arc<Mutex<`, as that data is immutable: an `Arc` should be enough to share this across the streams (but again, I need to check this).

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 7 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_join.rs:147]: Thanks for confirming! I was not really sure about this :)

File: rust/datafusion/src/physical_plan/hash_join.rs (hunk 9 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_join.rs]: @alamb thanks a lot for the review. Really great knowledge and notation about this. I changed the wording to match that.

My largest concern (which I think you also had) was that this PR was changing the `filter` code. I removed all those changes to have this PR focused on the `join`, and not force us to change the `filter` because we wanted this.

I agree with your comment on the other PR that we may want to keep the `filter` code as is. The change of making the `MutableDataArray` accepting multiple arrays makes the filter even slower, which IMO further strengths that idea.

**Comment by jorgecarleitao** [comment]
I've added a couple of more tests verifying that the parts of the join are correct. I promoted this PR out of a draft, as I am now confident that at least we would not merge buggy code ^_^

**Comment by andygrove** [comment]
I will be reviewing this tomorrow (Saturday). Thanks for all the work on this @jorgecarleitao

**Review by andygrove — APPROVED** [review]
LGTM. I think we should merge this, then get the DataFrame.join PR reviewed and merged, and then we can implement some of the TPC-H queries that have joins so we can run some comparisons with other query engines to see if that finds any issues.

**Review by alamb — COMMENTED** [review]
This is Epic 🥇  for @jorgecarleitao
